src/Workspaces/Core/MSBuild/Rpc/RpcClient.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild.Rpc;
+
+/// <summary>
+/// Implements the client side of the RPC channel used to communicate with the build host, which is using <see cref="RpcServer"/>.
+/// </summary>
+/// <remarks>
+/// The RPC system implemented here is pretty close to something like JSON-RPC; however since we need the Build Host to be usable in Source Build
+/// scenarios, we are limited to using only what is either in .NET or can be easily made buildable in Source Build. Thus existing solutions like StreamJsonRpc 
+/// are out. If at some point there is a standard RPC mechanism exposed in .NET or Source Build, we should delete this and use that instead.
+/// </remarks>
+internal sealed class RpcClient
+{
+    private readonly TextWriter _sendingStream;
+    private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+    private readonly TextReader _receivingStream;
+
+    private readonly ConcurrentDictionary<int, (TaskCompletionSource<object?>, System.Type? expectedReturnType)> _outstandingRequests = new();
+    private volatile int _nextRequestId = 0;
+
+    private readonly CancellationTokenSource _shutdownTokenSource = new CancellationTokenSource();
+
+    public RpcClient(Stream sendingStream, Stream receivingStream)
+    {
+        _sendingStream = new StreamWriter(sendingStream, JsonSettings.StreamEncoding);
+        _receivingStream = new StreamReader(receivingStream, JsonSettings.StreamEncoding);
+    }
+
+    public event EventHandler? Disconnected;
+
+    public void Start()
```

Comment:
would prefer async gets pushed out of this as well.

Reply:
In this case it wouldn't be useful, since the start/end surrounds the client calls. Think of this like https://learn.microsoft.com/en-us/dotnet/api/streamjsonrpc.jsonrpc.startlistening?view=streamjsonrpc-2.9 which isn't async either.